### PR TITLE
Preliminary refactor of quantum_variable.py

### PR DESCRIPTION
**Pull request by moar55**
## Description
This PR is a first step towards refactoring the quantum_variable file.
Mainly the goal of the PR is to do general cleanup of quantum_variable  such as
explicitly declaring instance/class attributes of `QuantumVariable` as well as
adding missing typings in various places(will be added in other places). 
Most importantly this PR refactors the `QuantumVariable`'s `__init__ ` method and
delegates the qv's name generation to the `QuantumSession`.

**Note**:  I have used the same approach for name generation in `duplicate` that I used in `__init__`. This means if `name` argument is provided a numerical suffix should be added if the name already exists in the `QuantumSession`. Now since the quantum session there is a fresh one, this should never happen. This assumption that the `name` argument must be unique to the `QuantumSession` was explicit earlier(an exception is raised otherwise), now it is not really enforced. Now it is not for convenience of re-using the name generation method. Not sure if this is the best approach, I could also have a concrete check that the name is not a duplicate and raise an exception otherwise...

## Type of Change

- [ ] Feature (new functionality)
- [x]  Change Request (modification of existing functionality) - Practically none but technically there is a slight change in logic for name generation in `duplicate`, see above note.
- [ ] Bug Fix
- [x] Refactoring (no behavior change)
- [ ] Performance improvement
- [ ] Documentation
- [ ] CI / Build

## Breaking Change?
- [ ] Yes
- [ ] No
- [x] Not sure - see above note.

## What was changed?

- Explicit instance/class attributes were added to `QuantumVariable` (`quantum_variable.py`)
- Typings added in the same file
- Some docstrings added
- Most importantly, name generation logic is now delegated to `QuantumSession`, and __init__ method is  more concise now.
- Necessary relevant changes were done in other files too

## How was it tested?
Some tests were ran locally. 
Will update this when CI passes.

## Checklist
- [x] My code follows the project's coding standards
- [x] I have performed a self-review
- [ ] I have added/updated tests (referencing issue Test-IDs)
- [ ] All tests pass locally and in CI
- [ ] I have updated the documentation
- [ ] I have added a changelog entry to `changelog-dev.rst`
- [ ] Breaking changes are documented with migration path

## Reviewer Notes

- One change that required multiple file change, is the change of the attribute
`usre_given_name` to `is_fixed_name`. The rationale for changing the name is
that this attribute is used to determine if the QuantumVariable can be renamed or not. 
The earlier name doesn't seem to convey that semantics, but rather conveys why
the boolean is set to True or False which can be confusing at conditions
involving the attribute. Let me know if you disagree with this part ofc.
-  I set `reg` in `quantum_variable` to `Any` (should be a `list[Qubit] |
DynamicQubitArray`) because it is not possible to see
the methods that `DynamicQubitArray` exposes at compile-time. Maybe we need to
extend the compile-time typing of `DynamicQubitArray` accordingly, but for now I
think we must use an Any.
-  I think in the next steps it might make sense to set the attributes set in
`register_qv` explicitly in the __init__ file (perhaps by making `register_qv`
return them) although I am not sure how well that will work out.

### CODE REVIEW

File: src/qrisp/core/quantum_variable.py [hunk 1]
```diff
@@ -1027,7 +1009,8 @@ def get_measurement(
             raise Exception("Tried to get measurement from deleted QuantumVariable")
 
         if self.size == 0:
-            return {"": 1.0}
+            # TODO: Create a const DecodedMeasurmentResult that describes this construct.
+            return {"": 1.0}  # pyright: ignore[reportReturnType]
```
Comment (moar55) [inline src/qrisp/core/quantum_variable.py:1013]: @PietropaoloFrisoni Thoughts about this?

Comment (PietropaoloFrisoni) [reply src/qrisp/core/quantum_variable.py:1013]: I think this makes sense (we might create another constructor for `DecodedMeasurementResult`, but it makes sense to leave a TODO here and do it in a follow-up PR).

For context, this `DecodedMeasurementResult` was introduced recently in the backend refactoring process

**Comment by moar55** [comment]
Next-follow up [PR](https://github.com/moar55/Qrisp/pull/1), will be opened against repo's `main` when this is merged.

**Comment by PietropaoloFrisoni** [comment]
Thank you @moar55!

Do you mind if I ask for a Copilot review before I take a closer look? I recently started using it for my own PRs, and it seems to be quite useful (although it lacks context and/or hallucinates in some cases)

**Comment by moar55** [comment]
> Thank you @moar55!
> 
> Do you mind if I ask for a Copilot review before I take a closer look? I recently started using it for my own PRs, and it seems to be quite useful (although it lacks context and/or hallucinates in some cases)

Sure :smile:

**Comment by purva-thakre** [comment]
> Next-follow up [PR](https://github.com/moar55/Qrisp/pull/1), will be opened against repo's `main` when this is merged.

@moar55 You could stack your new PR on top of this. I am on a mission to make everyone use it 😅. In my defense, this should make each PR review short and self-contained. 

https://docs.github.com/en/pull-requests/how-tos/stacked-pull-requests

**Comment by moar55** [comment]
> > Next-follow up [PR](https://github.com/moar55/Qrisp/pull/1), will be opened against repo's `main` when this is merged.
> 
> @moar55 You could stack your new PR on top of this. I am on a mission to make everyone use it 😅. In my defense, this should make each PR review short and self-contained.
> 
> https://docs.github.com/en/pull-requests/how-tos/stacked-pull-requests

I thought you can't stack prs from a fork :thinking:

**Comment by purva-thakre** [comment]
> I thought you can't stack prs from a fork 🤔

Thanks for the reminder! I forgot you are not a committer yet.

File: src/qrisp/core/quantum_variable.py [hunk 2]
```diff
@@ -470,12 +442,16 @@ def delete(self, verify=False, recompute=False):
...
         while i < len(QuantumVariable.live_qvs):
-            if QuantumVariable.live_qvs[i]() is None:
-                QuantumVariable.live_qvs.pop(i)
+            live_qv = QuantumVariable.live_qvs[i]()
+            if live_qv is None:
+                del QuantumVariable.live_qvs[i]
+                continue
+
+            if not live_qv.name:
                 continue
```
[the remark this answers is not in the document]
Comment (moar55) [reply src/qrisp/core/quantum_variable.py]: I change this whole loop in an upcoming PR, so I would rather not resolve this here right away.

File: src/qrisp/core/quantum_variable.py [hunk 3]
```diff
@@ -505,6 +484,14 @@ def duplicate(self, name=None, qs=None, init=False, qubits=None):
...
             If set to True, the :meth:`init_from <qrisp.QuantumVariable.init_from>`
             method of the result will be called on self. The default is False.
+        qubits: list[Qubit] | DynamicQubitArray, optional
+            Qubits to inialize the duplicated quantum variable from.
+            If not provided ``self.size`` is used for qubits extraction.
+
+        Raises
+        ------
+        QuantumVariableNamingError
+            Name of qv is already used in this QuantumSession.
```
[the remark this answers is not in the document]
Comment (moar55) [reply src/qrisp/core/quantum_variable.py]: I changed to a more general statement since merging can raise for other reasons...

File: tests/core_tests/test_variable_naming.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+* This Source Code may also be made available under the following Secondary
+* Licenses when the conditions for such availability set forth in the Eclipse
+* Public License, v. 2.0 are satisfied: GNU General Public License, version 2
+* with the GNU Classpath Exception which is
+* available at https://www.gnu.org/software/classpath/license.html.
+*
+* SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+********************************************************************************
+"""
+
```
Comment (PietropaoloFrisoni) [inline tests/core_tests/test_variable_naming.py:17]:
```suggestion
# ********************************************************************************
# * Copyright (c) 2026 the Qrisp authors
# *
# * This program and the accompanying materials are made available under the
# * terms of the Eclipse Public License 2.0 which is available at
# * http://www.eclipse.org/legal/epl-2.0.
# *
# * This Source Code may also be made available under the following Secondary
# * Licenses when the conditions for such availability set forth in the Eclipse
# * Public License, v. 2.0 are satisfied: GNU General Public License, version 2
# * with the GNU Classpath Exception which is
# * available at https://www.gnu.org/software/classpath/license.html.
# *
# * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
# ********************************************************************************

```

File: src/qrisp/core/quantum_variable.py [hunk 5]
```diff
@@ -230,124 +277,54 @@ def __init__(self, size, qs=None, name=None):
...
-                raise Exception("Tried to trace Qrisp code using make_jaxpr (use make_jaspr instead)")
+        from qrisp.core.quantum_session import QuantumSession
 
-            self.qubit_cache = {}
-        elif qs is not None:
-            self.qs = qs
+        if isinstance(self.qs, QuantumSession):
+            declaration_stack_level = 1 if type(self) is QuantumVariable else 2
+            (self.name, self.is_fixed_name) = self.qs.generate_name(name, self, declaration_stack_level + 1)
         else:
```
Comment (PietropaoloFrisoni) [inline src/qrisp/core/quantum_variable.py:308]: could this be a single `elif name is None` directive? I believe `ruff` is flagging the same

File: src/qrisp/core/quantum_variable.py [hunk 6]
```diff
@@ -540,39 +526,25 @@ def duplicate(self, name=None, qs=None, init=False, qubits=None):
...
+            (duplicate.name, duplicate.is_fixed_name) = new_qs.generate_name(
+                name if name is not None else self.name,
+                duplicate,
+                declaration_stack_level,
+                name is None,
+            )
         else:
-            duplicate.user_given_name = False
+            duplicate.name = name if name is not None else self.name + "_dupl"
+            duplicate.is_fixed_name = False
```
Comment (PietropaoloFrisoni) [inline src/qrisp/core/quantum_variable.py:540]: Should this mirror `__init__`'s handling? 

That is:

```
`is_fixed_name = not name.endswith("*")` / `name.removesuffix("*")`
```

File: src/qrisp/core/quantum_variable.py [hunk 7]
```diff
@@ -955,7 +927,10 @@ def get_measurement(
         filename=None,
         precompiled_qc=None,
```
Comment (PietropaoloFrisoni) [inline src/qrisp/core/quantum_variable.py:928]: `precompiled_qc` is missing in the docstring (this is a pre-existing problem since this function is a bit messy)

File: src/qrisp/core/quantum_variable.py [hunk 8]
```diff
@@ -505,6 +486,14 @@ def duplicate(self, name=None, qs=None, init=False, qubits=None):
         init : bool, optional
             If set to True, the :meth:`init_from <qrisp.QuantumVariable.init_from>`
             method of the result will be called on self. The default is False.
+        qubits: list[Qubit] | DynamicQubitArray, optional
+            Qubits to initialize the duplicated quantum variable from.
+            If not provided ``self.size`` is used for qubits extraction.
+
+        Raises
+        ------
+        Exception
```
Comment (PietropaoloFrisoni) [inline src/qrisp/core/quantum_variable.py:495]: Although AI technically considers this a breaking change, I think we should try to replace a generic `Exception` with a better error message class which eventually inherits from it (`RuntimeError`, `AttributeError`, `NotImplementedError`, etc.).

It must be said though that this can make the JAX tests blowup (a strange phenomenon I observed once and I never fully understood why), but this is a very rare case

File: src/qrisp/core/quantum_variable.py [hunk 9]
```diff
@@ -490,8 +468,11 @@ def is_deleted(self):
                 return True
         return False
 
-    def duplicate(self, name=None, qs=None, init=False, qubits=None):
-        r"""Duplicates the QuantumVariable in the sense that a new QuantumVariable is
+    def duplicate(self, name: str | None = None, qs=None, init=False, qubits=None):
```
Comment (PietropaoloFrisoni) [inline src/qrisp/core/quantum_variable.py:471]: Can we add type hints for the other parameters?

If you believe we should add more custom types to the `qrisp.typing` module, please let me know

File: src/qrisp/core/quantum_variable.py [hunk 10]
```diff
@@ -230,124 +277,54 @@ def __init__(self, size, qs=None, name=None):
...
-            # If this didn't work, generate a generic, unique name
-            if not name_found:
-                while True:
-                    try:
-                        self.name = self.get_unique_name()
-                        self.qs.register_qv(self, size)
-                        break
-                    except QuantumVariableNamingError:
-                        pass
+        from qrisp.jasp.tracing_logic import flatten_qv, unflatten_qv
```
Comment (PietropaoloFrisoni) [inline src/qrisp/core/quantum_variable.py:317]: Is this imported here because of a circular import issue? If yes, can we please leave a NOTE or comment? Might be useful when we will address this issue in a more general PR

File: src/qrisp/core/quantum_variable.py [hunk 11]
```diff
@@ -26,12 +29,43 @@
 from qrisp.core.compilation import qompiler
 
 if TYPE_CHECKING:
+    from qrisp.core.quantum_session import QuantumSession
     from qrisp.interface.measurement_result import DecodedMeasurementResult
+    from qrisp.jasp import TracingQuantumSession
+
+
+def extract_quantum_session(qs: QuantumSession | None) -> QuantumSession | TracingQuantumSession:
```
Comment (PietropaoloFrisoni) [inline src/qrisp/core/quantum_variable.py:37]: Yes, we definitely need to resolve this entanglement issue with the JAX compilation pipeline.

Right now we have a lot of functions that return completely different objects at runtime depending on whether we are tracing or not, which makes everything more difficult to maintain.

This is a long standing issue and of course we should not address it here, but thanks for raising the problem

File: src/qrisp/core/quantum_variable.py [hunk 12]
```diff
@@ -470,12 +447,13 @@ def delete(self, verify=False, recompute=False):
 
         i = 0
         while i < len(QuantumVariable.live_qvs):
-            if QuantumVariable.live_qvs[i]() is None:
-                QuantumVariable.live_qvs.pop(i)
+            live_qv = QuantumVariable.live_qvs[i]()
```
Comment (PietropaoloFrisoni) [inline src/qrisp/core/quantum_variable.py:450]: Thanks for the fix! Is this tested?

File: src/qrisp/core/quantum_session.py [hunk 13]
```diff
@@ -189,35 +195,146 @@ def __init__(self, backend=None):
...
         """
-        if qv.name in [temp_qv.name for temp_qv in self.qv_list + self.deleted_qv_list]:
-            raise QuantumVariableNamingError("Variable name " + str(qv.name) + " already exists in quantum session")
+        if not self._is_fresh_name(qv.name):
+            raise QuantumVariableNamingError(f"Variable name {qv.name} already exists in quantum session")
 
-        # Hand qubits to quantum variable
-        qv.reg = self.request_qubits(size, name=qv.name)
+        # Hand qubits to quantum variable if size is provided.
+        if size is not None:
```
Comment (PietropaoloFrisoni) [inline src/qrisp/core/quantum_session.py:330]: Thanks for the fix (now we should have parity between `TracingQuantumSession` and `QuantumSession`). Is this already tested?

File: src/qrisp/core/quantum_session.py [hunk 14]
```diff
@@ -31,7 +33,11 @@
 )
 from qrisp.core.quantum_variable import QuantumVariable
 from qrisp.core.session_merging_tools import multi_session_merge
-from qrisp.misc import get_depth_dic
+from qrisp.misc import find_calling_line, get_depth_dic
+
+
+class QuantumVariableNamingError(Exception):
```
Comment (PietropaoloFrisoni) [inline src/qrisp/core/quantum_session.py:39]: It's a bit strange that we have such a custom error only for the quantum variable name (why not for many other cases)?. Still, this was pre-existing and certainly not a problem is this PR

File: src/qrisp/core/quantum_variable.py [hunk 15]
```diff
@@ -884,12 +856,12 @@ def extend(self, amount, position=-1):
...
             for i in range(amount):
-                insertion_qubits[i].identifier = (
-                    self.name + "_ext_" + str(self.qs.qubit_index_counter[0]) + "." + str(self.size)
+                insertion_qubits[i].identifier = (  # pyright: ignore
+                    self.name + "_ext_" + str(self.qs.qubit_index_counter[0]) + "." + str(self.size)  # pyright: ignore
                 )
                 self.reg.insert(position + i, insertion_qubits[i])
 
-    def reduce(self, qubits, verify=False):
+    def reduce(self, qubits: Any, verify=False):
```
Comment (PietropaoloFrisoni) [inline src/qrisp/core/quantum_variable.py:864]: I introduced a `QubitLike` type in the typing module precisely for such case(s). Can we use it here, or does `pyright`/`pylance` complain?

File: src/qrisp/core/quantum_variable.py [hunk 16]
```diff
@@ -1012,6 +991,11 @@ def get_measurement(
         {1.0: 0.5, 3.0: 0.5}
 
         """
+        from qrisp.jasp import TracingQuantumSession
```
Comment (PietropaoloFrisoni) [inline src/qrisp/core/quantum_variable.py:994]: Same comment as above regarding the nested import

File: src/qrisp/core/quantum_variable.py [hunk 17]
```diff
@@ -1012,6 +991,11 @@ def get_measurement(
...
+
+        if isinstance(self.qs, TracingQuantumSession):
+            raise Exception("Tried to get measurement of a QuantumVariable in tracing mode")
```
Comment (PietropaoloFrisoni) [inline src/qrisp/core/quantum_variable.py:997]: Same comment as above, I think here we should raise a more specific error than a generic `Exception`

File: src/qrisp/core/quantum_variable.py [hunk 18]
```diff
@@ -1307,6 +1387,12 @@ def uncompute(self, do_it=True, recompute=False):
 
 
         """
+
+        from qrisp.jasp import TracingQuantumSession
+
+        if isinstance(self.qs, TracingQuantumSession):
+            raise Exception("Tried to uncompute a QuantumVariable in tracing mode")
```
Comment (PietropaoloFrisoni) [inline src/qrisp/core/quantum_variable.py:1394]: same as above

File: tests/core_tests/test_variable_naming.py
Comment (PietropaoloFrisoni) [inline tests/core_tests/test_variable_naming.py:1]: In general I would like to keep the test files matching the files in the source code. That is, every test should be called `test_<name of the source file>`. But in this case I think it makes sense to have a test file just for naming. I think the issue is that these classes `QuantumVariable`, `QuantumCircuit`, etc. are too large, so a test file for each one would be way, way too large at present.

File: src/qrisp/core/quantum_variable.py
Comment (PietropaoloFrisoni) [inline src/qrisp/core/quantum_variable.py:989]: I tried to run this example but it's broken as the parameter is `compilation_kwargs` instead of `transpilation_kwargs`. Can we please fix this?

File: src/qrisp/core/quantum_variable.py [hunk 19]
```diff
@@ -26,12 +29,43 @@
...
 if TYPE_CHECKING:
+    from qrisp.core.quantum_session import QuantumSession
     from qrisp.interface.measurement_result import DecodedMeasurementResult
+    from qrisp.jasp import TracingQuantumSession
+
+
+def extract_quantum_session(qs: QuantumSession | None) -> QuantumSession | TracingQuantumSession:
+    """Determine the QuantumSession a new QuantumVariable should be registered in.
+
+    While tracing (e.g. within jasp), the active :class:`TracingQuantumSession`
```
Comment (PietropaoloFrisoni) [inline src/qrisp/core/quantum_variable.py:40]: I tried building the documentation but `:class:`TracingQuantumSession` is a broken link (the class is not part of the documentation). The same link appears below

**Review by PietropaoloFrisoni — COMMENTED** [review]
Thanks a lot @moar55 , this is great 🚀

File: tests/core_tests/test_variable_naming.py (hunk 4 above)
Comment (moar55) [reply tests/core_tests/test_variable_naming.py]: I don't get this suggestion.
Why should I put '#' at the start of each line.

File: src/qrisp/core/quantum_variable.py (hunk 6 above)
Comment (moar55) [reply src/qrisp/core/quantum_variable.py]: Yes good point, but `is_fixed_name = name is None or name.endswith("*")` *

File: src/qrisp/core/quantum_session.py (hunk 13 above)
Comment (moar55) [reply src/qrisp/core/quantum_session.py]: This should be handled in a diff pr with the design we agreed on offline.

File: src/qrisp/core/quantum_variable.py (hunk 9 above)
Comment (moar55) [reply src/qrisp/core/quantum_variable.py]: I handled this in a follow [pr](https://github.com/moar55/Qrisp/pull/2/changes#diff-f4aae05bc295e8f9598abf9d65c633ac896d435d64f326b7bf0c4ca79847e530), btw lemme know if you think those should be merged together.

File: src/qrisp/core/quantum_session.py (hunk 13 above)
Comment (moar55) [reply src/qrisp/core/quantum_session.py]: Also about the test, there are tests for register_qv with None in the tracing_quantum_session variant. But I didn't add an explicit one should I?

File: src/qrisp/core/quantum_variable.py (hunk 15 above)
Comment (moar55) [reply src/qrisp/core/quantum_variable.py]: This is only executed in static mode if i understand correctly and i typed it properly in the affermentioned [pr](https://github.com/moar55/Qrisp/pull/2)

File: src/qrisp/core/quantum_variable.py (hunk 12 above)
Comment (moar55) [reply src/qrisp/core/quantum_variable.py]: This change shouldn't affect logic, but I don't see tests covering `delete`. Should I add ones?

File: tests/core_tests/test_variable_naming.py
Comment (moar55) [reply tests/core_tests/test_variable_naming.py]: Yeah we could have further modules in the future...

File: src/qrisp/core/quantum_variable.py (hunk 10 above)
Comment (moar55) [reply src/qrisp/core/quantum_variable.py]: It is necessary for circular import indeed.
The __init__.py of tracing_logic imports tracing_quantum_session which in turn imports QuantumVariable.

File: src/qrisp/core/quantum_variable.py (hunk 8 above)
Comment (moar55) [reply src/qrisp/core/quantum_variable.py]: I agree, but in this case the exception is raised by merge so the change should be there which is out of the scope of this pr.
Also i noticed init_from could raise if `qubits` are passed to `duplicate` that are not fresh or if `init`  is set to true when `duplicate` is called in tracing mode. I am not sure if this is clear enough to the user :thinking:  Maybe we should add a big warn in the docstring in the arguments section?

File: src/qrisp/core/quantum_variable.py (hunk 16 above)
Comment (moar55) [reply src/qrisp/core/quantum_variable.py]: Necessary as explained above, but will add a comment.
